### PR TITLE
Web server integration

### PR DESCRIPTION
**Pull request by fteychene**
## What this PR does

Add a web server on run to serve custom responses based on configuration loaded at the begining of the program

Crate : [hyper](https://github.com/hyperium/hyper)

## Related

Closes : #7 
Closes: #23

### CODE REVIEW

File: src/web.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+use hyper::service::{make_service_fn, service_fn};
+use hyper::{Body, Method, Request, Response, Server, StatusCode};
+
+use anyhow::{anyhow, Context, Error};
+
+use crate::expectation::model::{Expectation, HttpMethod};
+use crate::mock::{look_for_expectation, IncomingRequest};
+
+impl TryInto<HttpMethod> for &Method {
```
Comment (rlecomte) [inline src/web.rs:13]: ```**Library authors should usually not directly implement this trait**, but should prefer implementing the TryFrom trait, which offers greater flexibility and provides an equivalent TryInto implementation for free, thanks to a blanket implementation in the standard library. For more information on this, see the documentation for Into.```

File: src/web.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        }
+    }
+}
+
+impl TryInto<Response<Body>> for &Expectation {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> Result<Response<Body>, Error> {
+        Response::builder()
+            .status(self.response.status_code.unwrap_or(200))
```
Comment (rlecomte) [inline src/web.rs:33]: Shouldn't we force user to define response's status code?

File: src/web.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            .context(format!("Error creating http response for {:?}", self))
+    }
+}
+
+pub struct State {
+    pub expectations: Vec<Expectation>,
+}
+
+async fn handler<T>(req: Request<T>, state: Arc<RwLock<State>>) -> Result<Response<Body>, Error> {
+    let incoming_request = req.method().try_into().map(|method| IncomingRequest {
```
Comment (rlecomte) [inline src/web.rs:50]: I don't know what's the state of the art of error handling in rust but IMO `try_into` isn't a really clear way to perform.

File: src/web.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            )
+            .context(format!("Error creating http response for {:?}", self))
+    }
+}
+
+pub struct State {
+    pub expectations: Vec<Expectation>,
+}
+
+async fn handler<T>(req: Request<T>, state: Arc<RwLock<State>>) -> Result<Response<Body>, Error> {
```
Comment (rlecomte) [inline src/web.rs:49]: Could you let some comments in the code below? Adding some space would be pretty as well.

File: src/web.rs (hunk 2 above)
Comment (fteychene) [reply src/web.rs:33]: We could.
I guessed that by default if not provded, the satus should be a 200 status code.
THere are some design question about expectation that we should check after getting the first usable version.

File: src/web.rs (hunk 3 above)
Comment (fteychene) [reply src/web.rs:50]: Going to change around `TryFrom` but we have a limitation to go for a `Into` or a `From`.

If we implement `impl From<Method> for Option<HttpMethod>` as `Method` and `Option` aren't from this crate the compilation will fail so we have to manage the fact that HttpMethod doesn't modelize all ths hyper avaiable in method.

**Comment by rlecomte** [comment]
LGTM

**Pull request merged**